add jsonbytesdecoder to save_json function.

perosb opened the pull request:
Adds json bytes decoder to save_json.

## Description:
Regression when porting over to core json functions.

**Related issue (if applicable):** fixes #10751


## Checklist:
If the code does not interact with devices:
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

perosb committed:
add jsonbytesdecoder to save_json function. needed at least for notify.html5 to work

perosb committed:
don't test for bytes here

perosb committed:
remove unsed class; moved to util json

balloob commented on homeassistant/components/notify/html5.py:
```diff
@@ -133,17 +133,6 @@ def _load_config(filename):
     return {}
 
 
-class JSONBytesDecoder(json.JSONEncoder):
```
when is this needed and why did only HTML5 use it? What are the side effects of it being active?

perosb replied:
When subscribing push notification, chrome sends json data as bytes. Python default encoder doesn't support bytes.
https://docs.python.org/3/library/json.html#json.JSONEncoder

This encoder just checks if instance is bytes, if so, decode it.
On the above link there is a similar example for Complex types.

balloob replied:
I don't understand how that works 🤔 so this line is returning an object that is bytes? https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/notify/html5.py#L162

As far as I know, there is no bytes type in JSON.

I also think that we should just convert that type to string instead of updating the function that is used all over Home Assistant to support it.

andrey-git committed:
Add useragent-based detection of JS version (#10776)

* Add useragent-based detection of JS version

* Keep es5 as default meanwhile

* Update test

perosb commented:
When subscribing push notification, chrome sends json data as bytes. Python default encoder doesn't support bytes.
https://docs.python.org/3/library/json.html#json.JSONEncoder

This encoder just checks if instance is bytes, if so, decode it.
On the above link there is a similar example for Complex types.

perosb committed:
add support for multiple execution per execute request (#10844)

danielperna84 committed:
Upgrade HomeMatic, add devices (#10845)

tinloaf committed:
Climate component: add supported_features (#10658)

* Implement supported_features for the climate component

* Test supported features

* Convert generic thermostat to supported features

* Max / min temperature are not features

* Fix lint

* Min / max humidity are not features

* Linting

* Remove current temperature / humidity

* Move c-hacker-style constants to boring integers. Booo!

* Refactor all the climate platforms to use the new supported_features

* Force all climate platforms to implement supported_features

* Fix mistakes

* Adapt hive platform

* Move flags into a constant

* Calm the hound

fabaff committed:
Upgrade mutagen to 1.39 (#10851)

bachya committed:
Updated codeowner for Tile device tracker (#10861)

Julius2342 committed:
Revert "KNX: Added config option for broadcasting current time to KNX bus. (#10654)" (#10874)

This reverts commit cadd797200e339e165dbeff152169e4f4d716e2d.

As discussed within #10708 we should chose a different implementation. Therefore we should revert this change to avoid a breaking change.

fabaff committed:
Upgrade distro to 1.1.0 (#10850)

pvizeli committed:
Bugfix trigger state with multible entities (#10857)

* Bugfix trigger state with multible entities

* Fix numeric state

* fix lint

* fix dict

* fix unsub

* fix logic

* fix name

* fix new logic

* add test for state

* add numeric state test for unsub

* add test for multible entities

* Update numeric_state.py

* Update numeric_state.py

* Update state.py

* Fix logic for triple match

* Add clear to numeric state

* clear for state trigger

pvizeli committed:
tellstick fix DEPENDENCIES and update tellcore-net (#10859)

* Update requirements_all.txt

* Update tellstick.py

* Fix DEPENDENCIES

* Update requirements_all.txt

* fix format

* fix lint

* fix lint

* Update tellstick.py

* update tellcore-net

* update tellcore-net

* besser validate

balloob committed:
Update frontend to 20171130.0

fabaff committed:
Upgrade aiohttp to 2.3.5 (#10889)

fabaff committed:
Upgrade fastdotcom to 0.0.3 (#10886)

fabaff committed:
Upgrade schiene to 0.19 (#10887)

rytilahti committed:
Xiaomi Vacuum: remove deprecated calls (#10839)

* vacuum.xiaomi_miio: read dnd status properly instead of using imprecise dnd flag from vacuum_state

* vacuum.xiaomi_miio: use miio package instead of mirobo

* check only that wanted calls have taken place, ignore order of calls

* Fix linting issues

* Remove empty line after docstring

PhracturedBlue committed:
Create ecobee weather platform (#10869)

* Create ecobee weather component

* Update requirements_all for ecobee

* Fix missed lint issue

jeroenterheerdt committed:
Microsoft Text-to-speech: Fixing missing en-gb support bug (#10429)

* Fixing missing en-gb support bug

* Microsoft TTS adding support for rate, volume, pitch and contour.

* Microsoft TTS fixing support for jp-jp.

* Fixing linting error on line 67

* make impossible things possible :tada:

fabaff committed:
Upgrade youtube_dl to 2017.11.26 (#10890)

fabaff committed:
Upgrade yarl to 0.15.0 (#10888)

pvizeli committed:
Fix tests (#10891)

GenericStudent committed:
Refactored WHOIS sensor to resolve assumed key errors (#10662)

* Refactored WHOIS sensor to resolve assumed key errors

Altered it to now set an attribute key and value only if the attribute
is present in the WHOIS response. This prevents assumed keys (registrar)
from raising a KeyError on WHOIS lookups that don't contain registrar
information (onet.pl, wp.pl, for example).

* Removed non-used self._data

* WHOIS sensor now creates a new local attributes dict and overrides

* Corrected typos, refactored error cases to clear state adn attributes

* Resolved double return and refactored error logging

andrey-git committed:
Serve latest extra_html in dev mode (#10863)

andrey-git committed:
Reload groups after saving a change via config API (#10877)

fabaff committed:
Version bump to 0.59.0

PhracturedBlue committed:
Update ecobee version to fix stack-trace issue (#10894)

raymccarthy committed:
Pybotvac multi (#10843)

* Update requirements_all.txt

* Update neato.py

PhracturedBlue committed:
Fix issues from review of ecobee weather component (#10903)

* Fix issues from review

* Don't use STATE_UNKNOWN

PhracturedBlue committed:
Fix issues from review of ecobee weather component (#10903)

* Fix issues from review

* Don't use STATE_UNKNOWN

turbokongen committed:
Bugfix #10902 (#10904)

turbokongen committed:
Bugfix #10902 (#10904)

fabaff committed:
Merge pull request #10898 from home-assistant/release-0-59

0.59

fabaff committed:
Merge branch 'master' into dev

bonzini committed:
More declarative timeout syntax for manual alarm control panel. (#10738)

More declarative timeout syntax for manual alarm control panel

nickovs committed:
Unpacking RESTful sensor JSON results into attributes. (#10753)

* Added support for extracting JSON attributes from RESTful values

Setting the json_attributes configuration option to true on the
RESTful sensor will cause the result of the REST request to be parsed
as a JSON string and if successful the resulting dictionary will be
used for the attributes of the sensor.

* Added support for extracting JSON attributes from RESTful values

Setting the json_attributes configuration option to true on the
RESTful sensor will cause the result of the REST request to be parsed
as a JSON string and if successful the resulting dictionary will be
used for the attributes of the sensor.

* Added requirement that RESTful JSON results used as attributes must be
objects, not lists.

* Expanded test coverage to test REFTful JSON attributes with and
without a value template.

* Added support for extracting JSON attributes from RESTful values

Setting the json_attributes configuration option to true on the
RESTful sensor will cause the result of the REST request to be parsed
as a JSON string and if successful the resulting dictionary will be
used for the attributes of the sensor.

* Added requirement that RESTful JSON results used as attributes must be
objects, not lists.

* Expanded test coverage to test REFTful JSON attributes with and
without a value template.

* sensor.envirophat: add missing requirement (#7451)

Adding requirements that is not explicitly pulled in by the library
that manages the Enviro pHAT.

* PyPI Openzwave (#7415)

* Remove default zwave config path

PYOZW now has much more comprehensive default handling for the config
path (in src-lib/libopenzwave/libopenzwave.pyx:getConfig()). It looks in
the same place we were looking, plus _many_ more. It will certainly do a
much better job of finding the config files than we will (and will be
updated as the library is changed, so we don't end up chasing it). The
getConfig() method has been there for a while, but was subsntially
improved recently.

This change simply leaves the config_path as None if it is not
specified, which will trigger the default handling in PYOZW.

* Install python-openzwave from PyPI

As of version 0.4, python-openzwave supports installation from PyPI,
which means we can use our 'normal' dependency management tooling to
install it. Yay.

This uses the default 'embed' build (which goes and downloads
statically sources to avoid having to compile anything locally). Check
out the python-openzwave readme for more details.

* Add python-openzwave deps to .travis.yml

Python OpenZwave require the libudev headers to build. This adds the
libudev-dev package to Travis runs via the 'apt' addon for Travis.

Thanks to @MartinHjelmare for this fix.

* Update docker build for PyPI openzwave

Now that PYOZW can be install from PyPI, the docker image build process
can be simplified to remove the explicit compilation of PYOZW.

* Add datadog component (#7158)

* Add datadog component

* Improve test_invalid_config datadog test

* Use assert_setup_component for test setup

* Fix object type for default KNX port

#7429 describes a TypeError that is raised if the port is omitted in the config for the KNX component (integer is required (got type str)). This commit changes the default port from a string to an integer. I expect this will resolve that issue...

* Added support for extracting JSON attributes from RESTful values

Setting the json_attributes configuration option to true on the
RESTful sensor will cause the result of the REST request to be parsed
as a JSON string and if successful the resulting dictionary will be
used for the attributes of the sensor.

* Added requirement that RESTful JSON results used as attributes must be
objects, not lists.

* Expanded test coverage to test REFTful JSON attributes with and
without a value template.

* Added support for extracting JSON attributes from RESTful values

Setting the json_attributes configuration option to true on the
RESTful sensor will cause the result of the REST request to be parsed
as a JSON string and if successful the resulting dictionary will be
used for the attributes of the sensor.

* Added requirement that RESTful JSON results used as attributes must be
objects, not lists.

* Expanded test coverage to test REFTful JSON attributes with and
without a value template.

* Added support for extracting JSON attributes from RESTful values

Setting the json_attributes configuration option to true on the
RESTful sensor will cause the result of the REST request to be parsed
as a JSON string and if successful the resulting dictionary will be
used for the attributes of the sensor.

* Added requirement that RESTful JSON results used as attributes must be
objects, not lists.

* Expanded test coverage to test REFTful JSON attributes with and
without a value template.

* Fixed breaks cause by manual upstream merge.

* Added one extra blank line to make PyLint happy.

* Switched json_attributes to be a list of keys rather than a boolean.

The value of json_attributes can now be either a comma sepaated list
of key names or a YAML list of key names. Only matching keys in a
retuned JSON dictionary will be mapped to sensor attributes.

Updated test cases to handle json_attributes being a list.

Also fixed two minor issues arrising from manual merge with 0.58 master.

* Added an explicit default value to the json_attributes config entry.

* Removed self.update() from __init__() body.

* Expended unit tests for error cases of json_attributes processing.

* Align quotes

fabaff committed:
Bump dev to 0.60.0.dev0 (#10912)

molobrakos committed:
Update eliqonline.py (#10914)

Channel id is now required (change in API)

RyuzakiKK committed:
Add iAlarm support (#10878)

* Add iAlarm support

* Minor fixes to iAlarm

* Rename ialarmpanel to ialarm and add a check for the host value

* corrections in the value validation of ialarm

* add a missing period on ialarm

perosb committed:
Merge branch 'save_json-bytes-decode' into dev

perosb committed:
move back bytes decoding to html5 and only do it there
